### PR TITLE
Refactor communicator tests

### PR DESCRIPTION
**Pull request by ryan-jacobs**
This is a follow-up to Issue #7 that surfaced in the related PR #8.

In #8 some refactoring was required, which also led to test modifications. While addressing this we discovered that several tests that provide coverage for VuFindHarvest\OaiPmh\Communicator were problematically declared inside HarvesterFactoryTest. HarvesterFactoryTest was serving as a basis for these tests as new Harvester objects originally made Communicator calls via their constructor. PR #8 removes those constructor calls, and thus breaks the related tests. This follow addresses this by:

1. Moving tests that center on Communicator logic from HarvesterFactoryTest into a new CommunicatorTest class.
2. Restoring Communicator coverage temporarily removed during #8.

While this is a direct follow-up to #8 it is not strictly dependent on that PR. The new Communicator tests here should pass both before and after a merge on #8.

### CODE REVIEW

**Comment by ryan-jacobs** [comment]
Here's a snapshot of test coverage report **before** any merge on PR #8 (it's not clear to me if this can be retroactively generated). We assume coverage will drop after #8 is merged, but this PR should restore it.

Screen Shot 2022-09-26 at 7 05 13 PM

**Comment by demiankatz** [comment]
@ryan-jacobs, thanks for this! If you want to generate coverage reports on your own, you can just run `vendor/bin/phing phpunit` and the coverage files will be generated under /tmp/build/vufindharvest (as long as you have either xdebug or phpcov installed to collect the data). You can also set a Phing property to override the /tmp target if you need to.

It also looks like this one needs a php-cs-fixer run to clean up a minor whitespace issue.

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (demiankatz) [inline tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php]: Might want to make this 2022 since this is new.

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (demiankatz) [inline tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php:25]: Feel free to give yourself credit for this if you wish!

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (demiankatz) [inline tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php]: You should reverse these parameters -- first is "expected" and second is "actual".

Comment (demiankatz) [reply tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php]: Maybe it would be slightly more clear if you set `$expectedResponse = $this->getIdentifyResponse();` earlier in the test and use that instead of calling the method twice (but I'll leave that to your discretion).

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (demiankatz) [inline tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php]: If you wanted to be more specific, I suspect you could use `$this->once()` instead of `$this->any()` for the two counts in this test... I don't think either of those things should be called more than once.

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (demiankatz) [inline tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php]: Reversed parameters again.

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (demiankatz) [inline tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php]: I'm not sure if you really need to use a real response here, since the communicator doesn't care about the contents of the requests it's making. If you want to avoid the very long line, you could do something like `<mock />` and I expect the results would be the same.

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (demiankatz) [inline tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php:58]: It might make sense to add the ability to inject the Communicator with a mock \VuFindHarvest\ConsoleOutput\OutputInterface (using `setOutputWriter()`) so that you can set expectations about output ... e.g. to capture the logic that displays a "wait" message during the retry loop.

**Review by demiankatz — CHANGES_REQUESTED** [review]
A few other (mostly minor, and sometimes optional) suggestions. :-)

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (ryan-jacobs) [reply tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php:25]: Will do. FYI, institutional credit for this and #8 also goes to Center for Research Libraries (www.crl.edu) ... just an FYI in case your curious about org context 😄

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (ryan-jacobs) [reply tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php]: Whoops, a silly mistake there. Also, good point about establishing one variable.

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (ryan-jacobs) [reply tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php]: True. In https://github.com/vufind-org/vufindharvest/pull/10/commits/3e8884b82ce2b859c6ed422e14cdb6692ce5c96a I updated several `expects()` calls to have a more precise frequency. This actually revealed that the test503Retry case wasn't running to completion! I turns out that `getStatusCode()` is called **twice** per request iteration (thus **4** times total for the test), so an adjustment was needed there.

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (ryan-jacobs) [reply tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php:58]: Hummm, are you suggesting to do this inside `getCommunicator()`, but then access the mocked writer later in each test? It's not clear to me how I could do that in a generic way, but I may certainly be missing something simple about your suggestion. All that said, in https://github.com/vufind-org/vufindharvest/pull/10/commits/3e8884b82ce2b859c6ed422e14cdb6692ce5c96a I did add a test-specific version of this idea in test503Retry in order to check that the Console output line about the 503 response and delay was written.

**Comment by ryan-jacobs** [comment]
Thanks for the suggestions @demiankatz! I think the most recent updates cover everything, with the possible exception of the console writer test coverage.

File: tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php
Comment (demiankatz) [reply tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php:58]: There are a few ways you could approach this (e.g. create an $outputWriter property in the test class, and then set that to a mock writer in getCommunicator; or, add a second parameter to getCommunicator so you could preconfigure your output writer and inject it), but none of them are especially satisfying or clean-feeling, so I think addressing this on a test-by-test basis as you have makes sense here.

**Review by demiankatz — APPROVED** [review]
Everything is passing, so I'm merging this now. Thanks again, @ryan-jacobs!

**Pull request merged**
